### PR TITLE
Option to create a client that runs DNS resolutions for every new connection

### PR DESCRIPTION
**Pull request by idelpivnitskiy**
Motivation:

In scenarios when an application creates many clients, and there is a low rate of creating new connections, continuous background DNS resolutions can add too much load on DNS server.

Modifications:

- Introduce `HttpClients.DiscoveryStrategy` that allows users to change the resolutions strategy, with an option to  run ServiceDiscoverer on every new connection;
- Introduce `HttpClients` factories that take `DiscoveryStrategy` into account;
- Add similar factories for `GrpcClients`;
- Refactor `GlobalDnsServiceDiscoverer` to accommodate new behavior;
- Add `ResolvingConnectionFactoryFilter` that implement a new discovery strategy;
- Disable `retryServiceDiscoveryErrors` when it's known that SD will never fail;
- Don't allow users to override `ServiceDiscoverer` through the builder API when `DiscoveryStrategy.ON_NEW_CONNECTION` is used;
- Skip `RetryingServiceDiscoverer` and `StatusAwareServiceDiscoverer` when `HttpClients.NoRetriesStrategy` is set;
- Test new behavior;

Result:

Users can create a client that runs DNS resolutions every time the client opens a new connection.

### CODE REVIEW

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpClients.java [hunk 1]
```diff
@@ -175,6 +205,64 @@ public static SingleAddressHttpClientBuilder<HostAndPort, InetSocketAddress> for
...
+     * The returned builder can be customized using {@link SingleAddressHttpClientBuilderProvider}.
+     *
+     * @param address the {@code UnresolvedAddress} to connect to, resolved using a DNS {@link ServiceDiscoverer} for
+     * every new connection. This address will also be used for the {@link HttpHeaderNames#HOST} header value.
+     * Use {@link SingleAddressHttpClientBuilder#unresolvedAddressToHost(Function)} if you want to override that
+     * value or {@link SingleAddressHttpClientBuilder#hostHeaderFallback(boolean)} if you want to disable this behavior.
+     * @return new builder for the address
+     * @see SingleAddressHttpClientBuilderProvider
+     */
+    public static SingleAddressHttpClientBuilder<HostAndPort, InetSocketAddress> forSingleAddressResolveOnDemand(
```
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpClients.java:258]: Please help with a better name 🙈

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpClients.java [hunk 2]
```diff
@@ -175,6 +206,65 @@ public static SingleAddressHttpClientBuilder<HostAndPort, InetSocketAddress> for
...
+            final HostAndPort address) {
+        return applyProviders(address,
+                new DefaultSingleAddressHttpClientBuilder<>(address, unresolvedServiceDiscoverer()))
+                // Apply after providers to let them see these customizations.
+                .appendConnectionFactoryFilter(ResolvingConnectionFactoryFilter.INSTANCE)
+                .retryServiceDiscoveryErrors(NoRetriesStrategy.INSTANCE);
```
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpClients.java:265]: @daschl for context: this approach requires 3 (or at least 2) settings on the builder:
1. swapping default SD with `unresolvedServiceDiscoverer()`;
2. adding `ResolvingConnectionFactoryFilter`
3. (optional) disabling `retryServiceDiscoveryErrors` bcz retries are not required for `unresolvedServiceDiscoverer()`

I was trying to find a way how to simplify users' experience for configuring all required components with a single API call. As the result I introduced these new factories on `HttpClients`. 
I also considered something like `HttpClientFunctions` that will effectively be `UnaryOperator<SingleAddressHttpClientBuilder<HostAndPort, InetSocketAddress>>` and users will apply it to the builder when the want this mode, but found it a bit difficult for users to discover. 

I'm open to any other ideas how to make users experience better.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpClients.java [hunk 3]
```diff
@@ -175,6 +206,65 @@ public static SingleAddressHttpClientBuilder<HostAndPort, InetSocketAddress> for
...
+     * value or {@link SingleAddressHttpClientBuilder#hostHeaderFallback(boolean)} if you want to disable this behavior.
+     * @return new builder for the address
+     * @see SingleAddressHttpClientBuilderProvider
+     */
+    public static SingleAddressHttpClientBuilder<HostAndPort, InetSocketAddress> forSingleAddressResolveOnDemand(
+            final HostAndPort address) {
+        return applyProviders(address,
+                new DefaultSingleAddressHttpClientBuilder<>(address, unresolvedServiceDiscoverer()))
+                // Apply after providers to let them see these customizations.
+                .appendConnectionFactoryFilter(ResolvingConnectionFactoryFilter.INSTANCE)
```
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpClients.java:264]: Heads up: the position of this `ResolvingConnectionFactoryFilter` has some differences. If it's the first one in the chain, all subsequent filters will see already resolved address (like they do today with background resolutions). On the other hand, users of `TransportObserverConnectionFactoryFilter` won't be able to take DNS resolution time into account for opening a new connection. The only way for those users to properly measure time is to apply `TransportObserverConnectionFactoryFilter` via providers.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpClients.java (hunk 1 above)
Comment (daschl) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpClients.java:259]: Maybe if it is difficult to pick a name it can be worth exploring other options: we could reuse our current overloads and add one on the same name with an enum at the end i.e. (`ResolutionStrategy`) which could default to `POLL` but could be customized by the user to `ON_DEMAND`. That also leaves the door open to more customized strategies in the future if we ever need to.

Comment (daschl) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpClients.java:259]: Another option, but nor sure if feasible, would be to move it into the Builder itself as a configuration option?

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpClients.java (hunk 3 above)
Comment (daschl) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpClients.java:264]: Is this something we should document in the javadoc of the Builder / Factory somewhere?

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpClients.java (hunk 2 above)
Comment (daschl) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpClients.java:265]: I agree that wrapping this in a user-friendly way is good - see my additional comments on the potential other overload.

File: servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/HttpClientResolvesOnDemandTest.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import static io.servicetalk.transport.netty.internal.AddressUtils.serverHostAndPort;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+
+class HttpClientResolvesOnDemandTest {
+
+    @ParameterizedTest(name = "{displayName} [{index}]: protocol={0}")
+    @EnumSource(HttpProtocol.class)
+    void test(HttpProtocol protocol) throws Exception {
+        try (HttpServerContext serverContext = HttpServers.forAddress(localAddress(0))
```
Comment (daschl) [inline servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/HttpClientResolvesOnDemandTest.java:36]: similar question as on GRPC if this exercises the new code path.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ResolvingConnectionFactoryFilter.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+                return sd.discover(HostAndPort.of(address)).takeAtMost(1).firstOrError()
+                        .flatMap(resolvedAddresses -> {
+                            if (resolvedAddresses.size() > 1) {
+                                // In case DNS server returns multiple IPs, it's recommended to shuffle the result to
+                                // make sure the client balances load between all available IPs.
+                                List<ServiceDiscovererEvent<InetSocketAddress>> list =
+                                        resolvedAddresses instanceof List ?
+                                                (List<ServiceDiscovererEvent<InetSocketAddress>>) resolvedAddresses :
+                                                new ArrayList<>(resolvedAddresses);
+                                Collections.shuffle(list);
```
Comment (daschl) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ResolvingConnectionFactoryFilter.java:78]: From reading this I assume shuffling is the best thing we can do (compared to keeping state for a "true" round robin)?

File: servicetalk-grpc-netty/src/test/java/io/servicetalk/grpc/netty/GrpcClientResolvesOnDemandTest.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    void test() throws Exception {
+        String greetingPrefix = "Hello ";
+        String name = "foo";
+        String expectedResponse = greetingPrefix + name;
+        try (GrpcServerContext serverContext = GrpcServers.forAddress(localAddress(0))
+                .listenAndAwait((GreeterService) (ctx, request) ->
+                        succeeded(HelloReply.newBuilder().setMessage(greetingPrefix + request.getName()).build()));
+             BlockingGreeterClient client = GrpcClients.forAddressResolveOnDemand(serverHostAndPort(serverContext))
+                     .buildBlocking(new ClientFactory())) {
+            assertEquals(expectedResponse,
```
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-grpc-netty/src/test/java/io/servicetalk/grpc/netty/GrpcClientResolvesOnDemandTest.java:44]: ideally, always hamcrest because it generates better description messages in case of failures. but we are not always consistent. hope it will improve over-time.
I copied this test from `GrpcUdsTest` and modified without noticing it uses junit. I will adjust both tests

File: servicetalk-grpc-netty/src/test/java/io/servicetalk/grpc/netty/GrpcClientResolvesOnDemandTest.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+class GrpcClientResolvesOnDemandTest {
+
+    @Test
+    void test() throws Exception {
+        String greetingPrefix = "Hello ";
+        String name = "foo";
+        String expectedResponse = greetingPrefix + name;
+        try (GrpcServerContext serverContext = GrpcServers.forAddress(localAddress(0))
```
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-grpc-netty/src/test/java/io/servicetalk/grpc/netty/GrpcClientResolvesOnDemandTest.java:39]: For a new server, we always want to use `localAddress(0)` because it helps to avoid running on `0.0.0.0` address. That address might be restricted in some environments and cause test failures.

But I will modify the client factory to use "localhost" instead of taking hostname from `serverContext` for better clarity. Good call!

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ResolvingConnectionFactoryFilter.java (hunk 5 above)
Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ResolvingConnectionFactoryFilter.java:78]: Yes, our best effort to somehow distribute where the client connects to.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpClients.java (hunk 3 above)
Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpClients.java]: Actually, users will not be able to see it regardless of the order. I documented it as the last item on `DiscoveryStrategy.ON_NEW_CONNECTION` and took some notes to create follow-up tickets.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpClients.java (hunk 1 above)
Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpClients.java]: I really like the `enum` idea with overloads. It makes the API clear and gives a lot of freedom to document each strategy. Thanks a lot for the suggestion!

Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpClients.java]: Actually, your idea helped to solve another problem: how users can customize SD for this discovery mode. I made the necessary adjustments in https://github.com/apple/servicetalk/pull/2520/commits/c1779b28f78b592b250b5848208f5bae29511e47 and it's possible now 🎉 
Also helped to test all edge cases.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpClients.java [hunk 8]
```diff
@@ -335,4 +478,112 @@ public Completable closeAsyncGracefully() {
...
+        @Override
+        public Completable apply(final int i, final Throwable t) {
+            return Completable.failed(t);
+        }
+    }
+
+    /**
+     * Defines {@link ServiceDiscoverer} will be used.
+     */
+    public enum DiscoveryStrategy {
```
Comment (daschl) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpClients.java:522]: 👍 great explanation of each strategy.

File: servicetalk-grpc-netty/src/main/java/io/servicetalk/grpc/netty/GrpcClients.java [hunk 9]
```diff
@@ -90,6 +91,41 @@ public static GrpcClientBuilder<HostAndPort, InetSocketAddress> forAddress(final
...
+     * The returned builder can be customized using {@link GrpcClientBuilderProvider}.
+     *
+     * @param host host to connect to, resolved by default using a DNS {@link ServiceDiscoverer}.
+     * @param port port to connect to
+     * @param discoveryStrategy {@link DiscoveryStrategy} to use
+     * @return new builder for the address
+     * @see GrpcClientBuilderProvider
+     */
+    public static GrpcClientBuilder<HostAndPort, InetSocketAddress> forAddress(
+            final String host, final int port, final DiscoveryStrategy discoveryStrategy) {
```
Comment (daschl) [inline servicetalk-grpc-netty/src/main/java/io/servicetalk/grpc/netty/GrpcClients.java:107]: Consider moving the `DiscoveryStrategy` to a namespace where it is not tied to http or grpc, since it could be technically also used for other protocols?

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ResolvingConnectionFactoryFilter.java [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+                        .flatMap(resolvedAddresses -> {
+                            @Nullable
+                            ServiceDiscovererEvent<R> resolved;
+                            if (resolvedAddresses.isEmpty()) {
+                                resolved = null;
+                            } else if (resolvedAddresses.size() == 1) {
+                                resolved = resolvedAddresses instanceof List ?
+                                        ((List<ServiceDiscovererEvent<R>>) resolvedAddresses).get(0) :
+                                        resolvedAddresses.stream().findFirst().orElse(null);
+                                if (!AVAILABLE.equals(resolved.status())) {
```
Comment (daschl) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ResolvingConnectionFactoryFilter.java:86]: if resolved is not a list and nothing found (so orElse(null)) triggers (I know it's very unlikely) would this resolved.status() lead to a NPE?

**Review by daschl — APPROVED** [review]
looks good to me, just some minor comments to consider.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ResolvingConnectionFactoryFilter.java (hunk 10 above)
Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ResolvingConnectionFactoryFilter.java:86]: This path executes only when `resolvedAddresses.size() == 1` but I will add a check anyway to avoid the worst case scenario if there is a tricky collection impl.
I used `resolvedAddresses.stream().findFirst().orElse(null)` as a short way instead of getting an Iterator and checking `hasNext()`.
